### PR TITLE
feat: add concurrency safety

### LINKED ISSUE #58: Support concurrent merkel migrate invocations
Insert a pending migration entry into `merkel_meta` at the start of a migration and fail if one exists already.

### PR DESCRIPTION
**Pull request by eseliger**
Closes #58

### CODE REVIEW

File: src/adapter.ts [hunk 1]
```diff
@@ -26,9 +32,42 @@ export interface TableRow {
...
-    public abstract logMigrationTask(task: Task): Promise<void>
+    public abstract beginMigrationTask(task: Task): Promise<void>
+    public abstract finishMigrationTask(task: Task): Promise<void>
     public abstract checkIfTaskCanExecute(task: Task): Promise<void>
     public abstract close(): Promise<void>
+    protected abstract hasPendingMigration(): Promise<boolean>
+
+    public waitForPending(logger: Logger): Promise<boolean> {
+        let wasPending = false
+        return new Promise<boolean>(async (resolve, reject) => {
```
Comment (felixfbecker) [inline src/adapter.ts]: making the Promise resolver async will swallow any exception that happen within it

File: src/adapter.ts [hunk 2]
```diff
@@ -26,9 +32,42 @@ export interface TableRow {
...
+            // fail after 10 min
+            const timeout = setTimeout(() => reject(new PendingMigrationTimedOutError()), 1000 * 60 * 10)
```
Comment (felixfbecker) [inline src/adapter.ts]: If you just want to fail the Promise after 10min, I would use `Promise.race()` to race the timeout Promise against the operation Promise

File: src/adapters/postgres.ts [hunk 3]
```diff
@@ -58,30 +61,41 @@ export class PostgresAdapter extends DbAdapter {
...
         task.id = rows[0].id
     }
 
+    /**
+     * Marks the task as finished
+     */
+    public async finishMigrationTask(task: Task): Promise<void> {
+        const { rows } = await this.client.query(SQL`
+            UPDATE merkel_meta SET "applied_at" = ${task.appliedAt}, "head" = ${
+            task.head ? task.head.sha1 : null
```
Comment (felixfbecker) [inline src/adapters/postgres.ts]: Save these in variables so it doesn't get wrapped ugly?

File: src/cli.ts [hunk 4]
```diff
@@ -328,9 +351,16 @@ const migrationCommand = (type: TaskType) => async (argv: MigrationCommandArgv)
...
         const adapter = createAdapterFromUrl(argv.db!)
         await adapter.init()
+        await adapter.waitForPending(CLI_LOGGER)
         const head = await getHead()
-        for (const name of argv.migrations!) {
-            const task = new Task({ type, migration: new Migration(name) })
+        const tasks = await Promise.all(
+            argv.migrations!.map(async name => {
+                const task = new Task({ type, migration: new Migration(name), head })
+                await adapter.beginMigrationTask(task)
```
Comment (felixfbecker) [inline src/cli.ts]: Are these run in parallel intentionally? Won't they output log statements etc?

File: src/index.ts [hunk 5]
```diff
@@ -53,19 +53,41 @@ export class Status {
...
-            for (const task of commit.tasks) {
-                logger.log(task.toString() + '...')
-                /* istanbul ignore next */
-                const interval = setInterval(() => logger.log('.'), 100)
-                await task.execute(migrationDir, adapter, this.head, commit)
-                clearInterval(interval)
-                logger.log(' Success\n')
+        while (true) {
+            logger.log(this.toString())
+            const tasks = this.newCommits.reduce<Task[]>((prev, next) => prev.concat(next.tasks), [])
```
Comment (felixfbecker) [inline src/index.ts]: Why can't this be done anymore in a nested for loop?

File: src/test/e2e.test.ts [hunk 6]
```diff
@@ -55,7 +55,24 @@ describe('E2E', () => {
         await execFile('git', ['commit', '-m', `first migration\n\n[merkel up ${uuid}]`], { env: { PATH } })
         let status = await getStatus(adapter, await getHead())
         assert.equal(status.newCommits.length, 1)
-        await status.executePendingTasks('migrations', adapter, SILENT_LOGGER)
+        assert.equal(status.newCommits[0].tasks.length, 1)
+        // add pending task
+        await adapter.beginMigrationTask({ ...status.newCommits[0].tasks[0], head: await getHead() } as Task)
+        let shouldHaveFinished = false
+        // begin execution, this should wait
+        const promise = new Promise<void>(async (resolve, reject) => {
```
Comment (felixfbecker) [inline src/test/e2e.test.ts]: This swallows exceptions, you can make an async IIFE

File: src/index.ts (hunk 5 above)
Comment (eseliger) [reply src/index.ts]: it is below, but we need all of them before the migrations to create the pending entries

File: src/adapters/postgres.ts [hunk 7]
```diff
@@ -58,30 +61,45 @@ export class PostgresAdapter extends DbAdapter {
...
+     */
+    public async finishMigrationTask(task: Task): Promise<void> {
+        const head = task.head ? task.head.sha1 : null
+        const commit = task.commit ? task.commit.sha1 : null
+        const { rows } = await this.client.query(SQL`
+            UPDATE merkel_meta
+                SET "applied_at" = ${task.appliedAt},
+                    "head" = ${head},
+                    "commit" = ${commit}
+                WHERE "id" = ${task.id}
```
Comment (felixfbecker) [inline src/adapters/postgres.ts]: align statements like in other queries?

File: src/adapters/postgres.ts [hunk 8]
```diff
@@ -58,30 +61,45 @@ export class PostgresAdapter extends DbAdapter {
...
         task.id = rows[0].id
     }
 
+    /**
+     * Marks the task as finished
+     */
+    public async finishMigrationTask(task: Task): Promise<void> {
+        const head = task.head ? task.head.sha1 : null
+        const commit = task.commit ? task.commit.sha1 : null
+        const { rows } = await this.client.query(SQL`
```
Comment (felixfbecker) [inline src/adapters/postgres.ts]: `rows` is unused

File: src/adapter.ts [hunk 9]
```diff
@@ -26,9 +33,44 @@ export interface TableRow {
...
+    public waitForPending(logger: Logger): Promise<boolean> {
+        let wasPending = false
+        return Promise.race([
+            new Promise<never>((_, reject) =>
+                setTimeout(() => reject(new PendingMigrationTimedOutError()), 1000 * 60 * 10)
+            ),
+            (async (): Promise<boolean> => {
+                // fail after 10 min
+                let interval: NodeJS.Timer | undefined
+                while (true) {
```
[the remark this answers is not in the document]
Comment (eseliger) [reply src/adapter.ts]: there is only the return on l 60
don you think inf retries are too many? 😁

File: src/index.ts (hunk 5 above)
Comment (felixfbecker) [reply src/index.ts]: wouldn't that be possible with a nested for loop too?

File: src/adapter.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply src/adapter.ts]: But what in the case of the timeout? This loop would keep going

File: src/adapter.ts [hunk 10]
```diff
@@ -14,6 +16,11 @@ export class UnsupportedDialectError extends Error {
     }
 }
 
+export class PendingMigrationTimedOutError extends Error {
+    /* istanbul ignore next */
+    public name = 'PendingMigrationTimedOutError'
```
Comment (felixfbecker) [inline src/adapter.ts]: should be a string literal type and `readonly`

File: src/cli.ts [hunk 11]
```diff
@@ -283,33 +286,53 @@ yargs.command(
...
-            if (status.newCommits.some(commit => commit.tasks.length > 0)) {
-                if (argv.confirm) {
-                    const answer = await inquirer.prompt<{ continue: boolean }>({
-                        type: 'confirm',
-                        name: 'continue',
-                        message: 'Continue?',
-                    })
-                    if (!answer.continue) {
-                        process.exit(0)
+            while (true) {
```
Comment (felixfbecker) [inline src/cli.ts]: Should this retry an infinite amount of times?

File: src/cli.ts [hunk 12]
```diff
@@ -283,33 +286,53 @@ yargs.command(
...
+
+                    const hasChanged = await adapter.waitForPending(CLI_LOGGER)
+
+                    if (hasChanged) {
+                        process.stdout.write('The migrations have changed, reloading..\n\n')
+                        continue
+                    }
+                    // create pending tasks
+                    for (const task of tasks) {
+                        await adapter.beginMigrationTask(task)
```
Comment (felixfbecker) [inline src/cli.ts]: Between this and `waitForPendingMigrations()`, could another process have started the same migration?

File: src/cli.ts (hunk 11 above)
Comment (eseliger) [reply src/cli.ts]: hmm do you think it should fail?

File: src/cli.ts (hunk 12 above)
Comment (eseliger) [reply src/cli.ts]: hm yes, I added some failure recovery

**Pull request merged**